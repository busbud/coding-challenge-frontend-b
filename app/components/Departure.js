import React, { PropTypes } from 'react';
import classNames from 'classnames/bind';
import moment from 'moment';
import styles from '../styles/components/departure.css';
import { Row, Col } from 'react-flexbox-grid';

const cx = classNames.bind(styles);

const Departure = ({departure, destination, operator, origin}) => {
  console.log(operator);
  const departureTime = moment(departure.departure_time).format('h:mm a');
  const arrivalTime = moment(departure.arrival_time).format('h:mm a');

  console.log(departureTime);
  return (
    <div className={cx('departure-box')}>
      <Row>
        <Col xs={2}>
          <p>{ departureTime }</p>
          <i className="fa fa-arrow-down" aria-hidden="true"></i>
          <p>{ arrivalTime }</p>
        </Col>
        <Col xs={4}>
          <p>{destination.name}</p>
          <p>{origin.name}</p>

        </Col>
        <Col xs={3}>{operator.name}</Col>
        <Col xs={3}>Price</Col>
      </Row>
    </div>
  );
};

Departure.propTypes = {
  departure: PropTypes.object,
  operator: PropTypes.object,
  origin: PropTypes.object,
  destination: PropTypes.object
};

export default Departure;
