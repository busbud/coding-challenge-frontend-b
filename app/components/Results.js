var React = require('react');

class Results extends React.Component {
	render() {
		return (
			<div>Results</div>
		);
	}
}

module.exports = Results;