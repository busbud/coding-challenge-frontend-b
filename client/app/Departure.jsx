import React            from 'react'
import ReactTooltip     from 'react-tooltip'
import Icon             from './Icon.jsx'
import LoadingAnimation from './LoadingAnimation.jsx'

class Departure extends React.Component {

  constructor(props) {
    super(props);
    this.state = {};
    this.state.departure = this.props.data;
    this.formatPrice = this.formatPrice.bind(this);
  }
  
  formatPrice(number, currency) {
    return (number / 100).toLocaleString("en-US", {style:"currency", currency: this.props.config.currency});
  }

  render() {

    var dep = this.state.departure;
    
    if(dep) {

      let amenityIcons = Object.keys(dep.amenities)
        .filter(function(amenity) {
          return dep.amenities[amenity] === true;
        }.bind(this))
        .map(function(amenity, i) {
          return (
            <div 
              key={i} 
              data-tip={this.props.config.strings['AMENITY_'+amenity.toUpperCase()]} 
              data-for={'ToolTip'} 
              data-place='top'>
              <Icon src={'/images/icons/'+amenity+'.svg#Layer_1'}/>
            </div>
          );
        }.bind(this));

      return (
        <div className="departure" data-departureid={dep.id}>
          <div className="depatureInner">
        
            <div className="row">
            
              <div className="col-xs-6 col-sm-3">
                <div className="row">
                  <div className="col-xs-12 labelText">
                    {this.props.config.strings.PICKUP}
                  </div>
                  <div className="col-xs-12 time">
                    {dep.departure_display_time}
                  </div>
                  <div className="col-xs-12 location">
                    {dep.origin_location_name}
                  </div>
                </div>
              </div>
              
              <div className="col-xs-6 col-sm-3">
                <div className="row">
                  <div className="col-xs-12 labelText">
                    {this.props.config.strings.DROP_OFF}
                  </div>
                  <div className="col-xs-12 time">
                    {dep.arrival_display_time}
                  </div>
                  <div className="col-xs-12 location">
                    {dep.destination_location_name}
                  </div>
                </div>
              </div>
              
              <div className="col-xs-12 col-sm-3">
                <div className="row">
                  <div className="col-xs-12 labelText">
                    {dep.operator_display_name} {dep.amenities.display_name}
                  </div>
                </div>
                <div className="row">
                  <div className="col-xs-12 amenities">
                    {amenityIcons}
                  </div>
                </div>
              </div>

              <div className="col-xs-12 col-sm-3">
                <div className="row">
                  <div className="col-xs-6 col-sm-12 price">
                    {this.formatPrice(dep.prices.total)}<span className="currency">{this.props.config.currency} </span>
                  </div>
                  <div className="col-xs-6 col-sm-12">
                    <a className="button pull-right" href={dep.links.deeplink}>
                      {this.props.config.strings.BUY}
                    </a>
                  </div>
                </div>
              </div>
              
            </div>
          </div>
        </div>
      );      
    } else {

      return (
        <div className="departure loading">
          <LoadingAnimation />
        </div>
      )

    }
  }
}

export default Departure;