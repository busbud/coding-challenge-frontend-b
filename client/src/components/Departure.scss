@import "./variables.scss";

.departure {
  background-color: $lightGrey;
  color: $grey;
  width: 100%;
  box-shadow: 0 0 5px lighten($blue, 10%);
  border-radius: 5px;

  &__content {
    &__label {
      margin: 0.5rem 0;
      padding: 0.5rem 0;
      border-top: 2px solid $lightGrey;
      border-bottom: 2px solid $lightGrey;
      font-weight: bold;
      &:first-child {
        border-radius: 5px 5px 0 0;
        border-top: none;
      }
    }
    &__location {
      p {
        margin: 0.5em;
      }
    }
    &__duration {
      margin-top: 1em;
      padding-top: 0.5rem;
      border-top: 2px solid $lightGrey;
    }
  }

  &__reservation {
    button {
      padding: 0.5rem 2rem;
      width: 100%;
      margin-top: 1rem;
      font-size: 1.4rem;
      border: none;
      background-color: $blue;
      box-shadow: 0 0 5px lighten($blue, 10%);
      border-radius: 0 0 5px 5px;
      color: $lightGrey;
    }
  }

  &__footer,
  &__reservation__availability {
    display: none;
  }
}
