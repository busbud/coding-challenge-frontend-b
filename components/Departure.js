import PropTypes from "prop-types";
import format from "date-fns/format";

import { withTranslation } from "../i18n";
import { fonts } from "../theme";
import Button from "./Button";

const formatPrice = (price, currency) => {
  return `$${(price / 100).toFixed(2)} ${currency}`;
};

const formatTime = time => {
  return format(time, "HH:MM A");
};

const Departure = ({
  origin_city,
  origin_location,
  destination_city,
  destination_location,
  operator,
  departure_time,
  arrival_time,
  prices,
  t
}) => {
  return (
    <li>
      <div className="left">
        <div className="row">
          <p className="heading">{t("departs")}</p>
          <p>
            <strong>{formatTime(departure_time)}</strong> — {origin_city.name},{" "}
            {origin_location.name}
          </p>
        </div>
        <div className="row">
          <p className="heading">{t("arrives")}</p>
          <p>
            <strong>{formatTime(arrival_time)}</strong> —{" "}
            {destination_city.name}, {destination_location.name}
          </p>
        </div>
        <div className="row">
          <p className="heading">{t("operated-by")}</p> <p />
          <p>
            <strong>{operator.name}</strong>
          </p>
        </div>
      </div>
      <div className="right">
        <p className="price">{formatPrice(prices.total, prices.currency)}</p>
        <div className="button-wrapper">
          <Button>{t("buy-now")}</Button>
        </div>
      </div>
      <style jsx>{`
        li {
          background-color: white;
          border-radius: 8px;
          box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.24);
          display: grid;
          grid-template-columns: 1fr;
          grid-row-gap: 16px;
          grid-column-gap: 32px;
          padding: 20px;
        }
        @media (min-width: 640px) {
          li {
            grid-template-columns: 1fr 220px;
          }
        }
        .row {
          display: flex;
          flex-wrap: wrap;
        }
        .row + .row {
          margin-top: 16px;
        }
        @media (min-width: 800px) {
          .row {
            flex-wrap: nowrap;
          }
        }
        .heading {
          flex-basis: 100%;
          min-width: 150px;
          padding-top: 4px;
        }
        @media (min-width: 800px) {
          .heading {
            flex-basis: auto;
          }
        }
        p {
          margin: 0;
        }
        strong {
          font-family: ${fonts.slab};
          font-size: 1.25rem;
        }
        @media (min-width: 640px) {
          .right {
            align-items: flex-end;
            display: flex;
            flex-direction: column;
            justify-content: space-between;
          }
        }
        .price {
          font-family: ${fonts.slab};
          font-size: 2rem;
          margin: 0;
        }
        .button-wrapper {
          margin-top: 12px;
        }
      `}</style>
    </li>
  );
};

Departure.propTypes = {
  origin_city: PropTypes.object,
  origin_location: PropTypes.object,
  destination_city: PropTypes.object,
  destination_location: PropTypes.object,
  operator: PropTypes.object,
  departure_time: PropTypes.string,
  arrival_time: PropTypes.string,
  prices: PropTypes.object,
  t: PropTypes.func.isRequired
};

export default withTranslation("common")(Departure);
