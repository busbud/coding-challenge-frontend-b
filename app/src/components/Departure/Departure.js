import React from "react";

import "./Departure.scss";

const Departure = () => <div id="departure">Departure</div>;

export default Departure;
