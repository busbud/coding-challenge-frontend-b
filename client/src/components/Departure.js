import React from "react";
import LocalizedStrings from "react-localization";
import "./Departure.scss";

let strings = new LocalizedStrings({
  en: {
    from: "From",
    to: "To",
    when: "When",
    book: "Book",
    duration: "Duration: ",
    complete: "Complete",
    seats: "Available seats:"
  },
  fr: {
    from: "Depuis",
    to: "Vers",
    when: "Date",
    book: "Réserver",
    duration: "Durée : ",
    complete: "Complet",
    seats: "Sièges disponibles :"
  }
});

const Departure = ({
  arrival_time,
  available_seats,
  complete,
  departure_time,
  duration,
  prices,
  seatClass,
  from,
  to,
  lang
}) => {
  strings.setLanguage(lang);
  return (
    <div className={`departure ${complete ? "departure--complete" : ""}`}>
      <div className="departure__content">
        <p className="departure__content__label">{strings.from}</p>
        <div className="departure__content__location">
          <p>{from}</p>
          <p>{new Date(departure_time).toLocaleString(lang)}</p>
        </div>
        <p className="departure__content__arrow">&rarr;</p>
        <p className="departure__content__label">{strings.to}</p>
        <div className="departure__content__location">
          <p>{to}</p>
          <p>{new Date(arrival_time).toLocaleString(lang)}</p>
        </div>
        <div className="departure__content__duration">
          <span>{strings.duration}</span>
          {duration}
        </div>
      </div>
      <div className="departure__reservation">
        <div className="departure__reservation__availability">
          <div>{seatClass}</div>
          <p>
            {complete
              ? strings.complete
              : `${strings.seats} ${available_seats}`}
          </p>
        </div>
        <div className="departure__reservation__price">
          {prices.total} {prices.currency}
        </div>
        <button>{strings.book}</button>
      </div>
    </div>
  );
};

export default Departure;
