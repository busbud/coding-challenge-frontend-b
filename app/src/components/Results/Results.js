import React from "react";

import DeparturesList from "./../DeparturesList/DeparturesList";

import "./Results.scss";

const Results = ({ list }) => (
  <div id="results">
    <DeparturesList list={list} />
  </div>
);

export default Results;
