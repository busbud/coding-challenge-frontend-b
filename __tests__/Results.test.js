import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";

import Results from "../components/Results";
import Departure from "../components/Departure";

import {
  cities,
  departures,
  locations,
  operators,
  selected_date
} from "../lib/testData";

jest.mock("react-i18next", () => ({
  withTranslation: () => Component => {
    Component.defaultProps = {
      ...Component.defaultProps,
      i18n: { language: "en" },
      t: key => key
    };
    return Component;
  }
}));

describe("<Results />", () => {
  it("renders and matches the snapshot", () => {
    const wrapper = shallow(
      <Results
        cities={cities}
        locations={locations}
        departures={departures}
        operators={operators}
        selected_date={selected_date}
      />
    );
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("displays the selected date in English", () => {
    const wrapper = shallow(
      <Results
        cities={cities}
        locations={locations}
        departures={departures}
        operators={operators}
        selected_date={selected_date}
      />
    );

    const selectedDate = wrapper.find("h2");
    expect(selectedDate.text()).toContain("Thursday 1st August 2019");
  });

  describe("when the language is set to FR", () => {
    it("displays the selected date in French", () => {
      const i18n = {
        language: "fr"
      };
      const wrapper = shallow(
        <Results
          cities={cities}
          locations={locations}
          departures={departures}
          operators={operators}
          selected_date={selected_date}
          i18n={i18n}
        />
      );

      const selectedDate = wrapper.find("h2");
      expect(selectedDate.text()).toContain("jeudi 1er ao√ªt 2019");
    });
  });

  it("renders three <Departure /> components", () => {
    const wrapper = shallow(
      <Results
        cities={cities}
        locations={locations}
        departures={departures}
        operators={operators}
        selected_date={selected_date}
      />
    );
    expect(wrapper.find(Departure).length).toBe(3);
    wrapper.find(Departure).forEach(departure => {
      const departureProps = departure.props();
      const matchingDeparture = departures.find(
        departure => departure.id === departureProps.id
      );
      expect(departureProps.departure_time).toBe(
        matchingDeparture.departure_time
      );
    });
  });
});
