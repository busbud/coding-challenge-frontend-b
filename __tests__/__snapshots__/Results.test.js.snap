// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Results /> renders and matches the snapshot 1`] = `
<section
  className="jsx-4199177593 wrapper"
>
  <h2
    className="jsx-4199177593"
  >
    results-for
     
    Thursday 1st August 2019
  </h2>
  <ul
    className="jsx-4199177593"
  >
    <Departure
      arrival_time="2019-08-01T08:30:00"
      departure_time="2019-08-01T01:30:00"
      destination_city={
        Object {
          "id": 1,
        }
      }
      destination_location={
        Object {
          "city_id": 1,
          "id": 1,
        }
      }
      i18n={
        Object {
          "language": "en",
        }
      }
      key="1"
      operator={
        Object {
          "id": 1,
        }
      }
      origin_city={
        Object {
          "id": 1,
        }
      }
      origin_location={
        Object {
          "city_id": 1,
          "id": 1,
        }
      }
      prices={
        Object {
          "currency": "CAD",
          "total": 5200,
        }
      }
      t={[Function]}
    />
    <Departure
      arrival_time="2019-08-01T08:30:00"
      departure_time="2019-08-01T01:30:00"
      destination_city={
        Object {
          "id": 1,
        }
      }
      destination_location={
        Object {
          "city_id": 1,
          "id": 1,
        }
      }
      i18n={
        Object {
          "language": "en",
        }
      }
      key="2"
      operator={
        Object {
          "id": 1,
        }
      }
      origin_city={
        Object {
          "id": 1,
        }
      }
      origin_location={
        Object {
          "city_id": 1,
          "id": 1,
        }
      }
      prices={
        Object {
          "currency": "CAD",
          "total": 5200,
        }
      }
      t={[Function]}
    />
    <Departure
      arrival_time="2019-08-01T08:30:00"
      departure_time="2019-08-01T01:30:00"
      destination_city={
        Object {
          "id": 1,
        }
      }
      destination_location={
        Object {
          "city_id": 1,
          "id": 1,
        }
      }
      i18n={
        Object {
          "language": "en",
        }
      }
      key="3"
      operator={
        Object {
          "id": 1,
        }
      }
      origin_city={
        Object {
          "id": 1,
        }
      }
      origin_location={
        Object {
          "city_id": 1,
          "id": 1,
        }
      }
      prices={
        Object {
          "currency": "CAD",
          "total": 5200,
        }
      }
      t={[Function]}
    />
  </ul>
  <JSXStyle
    dynamic={
      Array [
        "#243b52",
      ]
    }
    id="3450715267"
  >
    section.__jsx-style-dynamic-selector{padding-left:24px;padding-right:24px;}h2.__jsx-style-dynamic-selector{color:#243b52;margin:0;padding-top:32px;text-align:center;}ul.__jsx-style-dynamic-selector{display:grid;grid-template-rows:1fr;grid-gap:32px;list-style:none;margin:0 auto;max-width:960px;padding-left:0;padding-top:32px;}
  </JSXStyle>
</section>
`;
